data-model/src/main/java/io/micronaut/data/model/DefaultCursoredPageable.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2017-2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.model;
+
+import io.micronaut.core.annotation.Creator;
+import io.micronaut.core.annotation.Introspected;
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+
+import java.util.Objects;
+
+/**
+ * The default cursored pageable implementation.
+ *
+ * @param page The page.
+ * @param startCursor The cursor that is pointing to the start of the data.
+ *                    This cursor will be used for forward pagination.
+ * @param endCursor The cursor that is pointing to the end of the data.
+ *                  This cursor will be used for backward pagination
+ * @param isBackward Whether user requested for backward pagination.
+ * @param size The size of a page
+ * @param sort The sorting
+ * @param requestTotal Whether to request the total count
+ *
+ * @author Andriy Dmytruk
+ * @since 4.8.0
+ */
+@Introspected
+record DefaultCursoredPageable(
+    int size,
+    @Nullable
+    Cursor startCursor,
+    @Nullable
+    Cursor endCursor,
+    boolean isBackward,
+    int page,
+    Sort sort,
+    boolean requestTotal
+) implements CursoredPageable {
```

Comment:
We can have that.

There can be a cursor for each row. A cursor is just a collection of fields that we sort by. So if we sort by name and id, the cursor is just `[name, id]` for a particular row. Storing cursors for each point in data seems a little excessive - I currently only store current cursor and next cursor (created from the last item on page), since that is enough to support going to next and previous page. Should we store all cursors like in jakarta?

Reply:
That would be the best. In the future, we want to support Jakarta Data so we would need to map their implementation into ours.